File: lokma/geometry/volume_engine.py
"""VolumeEngineService — turns a real-world footprint area into a volume.

Pure, stateless geometry. Each food has a shape prior (prism / cylinder /
paraboloid / flat); the volume is the footprint area extruded by an assumed
height according to that prior. Height comes from calibration/depth upstream;
here it is a parameter.
"""

from __future__ import annotations

from lokma.core.models import VolumeEstimate


class VolumeEngineService:
    """Computes volume (cm³) from a real-world area (cm²) and a shape prior."""

    #: Assumed thickness for a thin/flat layer (e.g. pizza, carpaccio), in cm.
    FLAT_LAYER_CM: float = 0.8

    def estimate_volume(self, real_area_cm2: float, shape: str, height_cm: float) -> VolumeEstimate:
        shape_key = (shape or "prism").lower()

        if shape_key in ("prism", "cylinder"):
            volume = real_area_cm2 * height_cm
            effective_height = height_cm
        elif shape_key == "paraboloid":
            volume = 0.5 * real_area_cm2 * height_cm
            effective_height = height_cm
        elif shape_key == "flat":
            volume = real_area_cm2 * self.FLAT_LAYER_CM
            effective_height = self.FLAT_LAYER_CM
        else:  # unknown shape -> simple extrusion fallback
            volume = real_area_cm2 * height_cm
            effective_height = height_cm

        return VolumeEstimate(
            volume_cm3=volume,
            shape=shape_key,
            height_cm=effective_height,
            real_area_cm2=real_area_cm2,
            method=f"geometric:{shape_key}",
        )

File: ios/LokmaCore/Sources/LokmaCore/VolumeEngine.swift
// VolumeEngine — mirror of `lokma/geometry/volume_engine.py`.
//
// Pure, stateless geometry: a real-world footprint area extruded by an assumed
// height according to the food's shape prior.
import Foundation

public struct VolumeEngine: Sendable {
    /// Assumed thickness for a thin/flat layer (e.g. pizza, carpaccio), in cm.
    public static let flatLayerCm: Double = 0.8

    public init() {}

    public func estimateVolume(realAreaCm2: Double, shape: String, heightCm: Double) -> VolumeEstimate {
        let shapeKey = (shape.isEmpty ? "prism" : shape).lowercased()
        let volume: Double
        let effectiveHeight: Double

        switch shapeKey {
        case "prism", "cylinder":
            volume = realAreaCm2 * heightCm
            effectiveHeight = heightCm
        case "paraboloid":
            volume = 0.5 * realAreaCm2 * heightCm
            effectiveHeight = heightCm
        case "flat":
            volume = realAreaCm2 * Self.flatLayerCm
            effectiveHeight = Self.flatLayerCm
        default:  // unknown shape -> simple extrusion fallback
            volume = realAreaCm2 * heightCm
            effectiveHeight = heightCm
        }

        return VolumeEstimate(
            volumeCm3: volume, shape: shapeKey, heightCm: effectiveHeight,
            realAreaCm2: realAreaCm2, method: "geometric:\(shapeKey)"
        )
    }
}
